### DISCUSSION TITLE
Dynamic Python Function Generation from GraphQL Schema for API Requests

### DISCUSSION THREAD

**Discussion by b8511** [mirumee/ariadne#1134] [General]
I'm looking for a method to dynamically generate Python functions that correspond to the queries and mutations defined in a GraphQL schema. After obtaining the schema, I want a system that can automatically create these functions, so I can simply import the generated code into my Python project and call the functions with the necessary parameters.

The ideal solution would facilitate seamless interactions with the GraphQL API by abstracting the request construction process. It would allow for changes in the schema without the need for manual updates to the function definitions in the client code.

I haved tried ariadnegraphql code generator but I had to specify a query like this:

query MyQuery {
    getCar(plate_number: "14-77-ER") {
      addresses
      brand
      mongoObjId
    }
}
and it created an Client for this query with fix params.

Any advice on how I can better leverage the capabilities of Ariadne would be awesome!

**Comment by rafalp** [comment]
Hello!

This is exactly what [Ariadne Codegen](https://github.com/mirumee/ariadne-codegen) does. It takes schema, GraphQL queries and generates fully typed client class you import in your code to run queries.

**Reply by b8511** [reply]
Alright, maybe I'm not understanding what is the use for queries_path, cause I just put an example of one of my queries in the file that queries_path is pointing to
like so:
query MyQuery {
getCar(plate_number: "14-77-ER") {
addresses
brand
mongoObjId
}
}

and got client.py file like this:
`
# Generated by ariadne-codegen
# Source: queries.graphql

from typing import Any, Dict

from .async_base_client import AsyncBaseClient
from .my_query import MyQuery


def gql(q: str) -> str:
    return q


class Client(AsyncBaseClient):
    async def my_query(self, **kwargs: Any) -> MyQuery:
        query = gql(
            """
              query MyQuery {
                  getCar(plate_number: "14-77-ER") {
                    addresses
                    brand
                    mongoObjId
                  }
              }
            """
        )
        variables: Dict[str, object] = {}
        response = await self.execute(query=query, variables=variables, **kwargs)
        data = self.get_data(response)
        return MyQuery.model_validate(data)
`

I wanted to not have a fixed param in the plate_number

**Comment by rafalp** [comment]
You need to change your query to accept variable:

```graphql
query MyQuery($plate: String!) {
  getCar(plate_number: $plate) {
    addresses
    brand
    mongoObjId
  }
}
```

Then generated client will have a `my_query(plate: string)` method.

**Reply by b8511** [reply]
It work. Lets Go! Sry for my ignorance

**Reply by b8511** [reply]
Is there a way to make queries.graphql automatically?

**Reply by rafalp** [reply]
Nope, GraphQL is too open standard for this to make a sense.

But you can try implementing such solution yourself through a plugin.

**Reply by b8511** [reply]
alright, anyways thank you for your time
